include/adaflo/sharp_interface_util.h

```diff
@@ -0,0 +1,1108 @@
+    }
+
+  // step 2 communicate (TODO)
+
+  // step 3: convert quadrature points to a pair of cells and reference cell quadrature
+  // point
+  std::vector<std::tuple<Point<spacedim>, double, std::pair<int, int>>> info;
+
+  const std::vector<bool>                    marked_vertices;
+  const GridTools::Cache<spacedim, spacedim> cache(tria, mapping);
+  const double                               tolerance = 1e-10;
+  auto                                       cell_hint = tria.begin_active();
+
+  for (const auto &point_and_weight : locally_owned_surface_points)
+    {
+      const auto cell_and_reference_coordinate = GridTools::find_active_cell_around_point(
+        cache, point_and_weight.first, cell_hint, marked_vertices, tolerance);
+
+      cell_hint = cell_and_reference_coordinate.first;
+
+      info.emplace_back(
+        cell_and_reference_coordinate.second,
+        point_and_weight.second,
+        std::pair<int, int>(cell_and_reference_coordinate.first->level(),
+                            cell_and_reference_coordinate.first->index()));
+    }
+
+  // step 4: compress data structures
+  std::sort(info.begin(), info.end(), [](const auto &a, const auto &b) {
+    return std::get<2>(a) < std::get<2>(b);
+  });
+
+  std::vector<std::pair<int, int>> cells;
+  std::vector<unsigned int>        ptrs;
+  std::vector<double>              weights;
+  std::vector<Point<spacedim>>     points;
+
+  std::pair<int, int> dummy{-1, -1};
+
+  for (const auto &i : info)
+    {
+      if (dummy != std::get<2>(i))
+        {
+          dummy = std::get<2>(i);
+          cells.push_back(std::get<2>(i));
+          ptrs.push_back(weights.size());
+        }
+      weights.push_back(std::get<1>(i));
+      points.push_back(std::get<0>(i));
+    }
+  ptrs.push_back(weights.size());
+
+  return {cells, ptrs, weights, points};
+}
+
+
+
+template <int dim, typename VectorType, typename BlockVectorType>
+void
+compute_force_vector_sharp_interface(const Triangulation<dim - 1, dim> &surface_mesh,
+                                     const Mapping<dim - 1, dim> &      surface_mapping,
+                                     const FiniteElement<dim - 1, dim> &surface_fe,
+                                     const Quadrature<dim - 1> &        surface_quad,
+                                     const Mapping<dim> &               mapping,
+                                     const DoFHandler<dim> &            dof_handler,
+                                     const DoFHandler<dim> &            dof_handler_dim,
+                                     const BlockVectorType &normal_vector_field,
+                                     const VectorType &     curvature_solution,
+                                     VectorType &           force_vector)
+{
+  // step 1) collect all locally-relevant surface quadrature points (cell,
+  // reference-cell position,
+  //  quadrature weight)
+  const auto [cells, ptrs, weights, points] =
```

Comment:
Do we require C++17 in adaflo already?

Reply:
I would hope so. In the other PR, I am using `std::filesystem` which is quite a nice C++17 feature.